### PR TITLE
🪟🎉 Connector form: Reload config if updated during connection check

### LINKED ISSUE #21041: Better handle control messages while editing the connector through UI (Part 2: UI update)
After https://github.com/airbytehq/airbyte/issues/20913 is complete, this story is to update the front-end to re-hydrate the connector setup form if the `check_for_update` API response indicated that there was a configuration change during that check (e.g. an oAuth token has changed).  See https://github.com/airbytehq/airbyte/issues/20913 for a full description.

image

From https://github.com/airbytehq/airbyte/issues/20913#issuecomment-1369563163: 

> Instead of returning the updated config in the response of the check_for_update call, what about sending the flag whether it updated automatically and have the frontend re-fetch the values of the form and repopulate it by calling (sources|destinations)/get (instead of showing a message to the user to reload manually)? This way no further backend changes would be needed and this kind of change is relatively simple in the frontend with our setup. This way the additional complexity of the connector updating its config during check is completely transparent to the user.

## Acceptance criteria

* If the configuration is updated on the "testing connection" call, the form is updated with the most recent data as if the user had refreshed the page

### PR DESCRIPTION
**Pull request by flash1293**
## What
Fixes https://github.com/airbytehq/airbyte/issues/21041 

This PR will make sure updates to the configuration of an existing source or destination during the connection check won't be lost due to a subsequent update dispatched by the UI.

## How

The flag indicating that the config got updated during the check is already returned to the webapp (`jobInfo.connectorConfigurationUpdated`). In `airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx` this response can be checked and if it's set, the actual `update` call can be prevented as it has been done implicitly by the connection check.

At this point the form should be reloaded with the new config. This is done by passing down a `reloadConfig` callback which is calling `queryClient.invalidateQueries` for the source/destination which will cause react-query to reload the config and pass down the new state.

## How to test

Unfortunately it's hard to cause a real update-on-check to happen - I tested by mocking the server like this (need to run `./gradlew airbyte-server:assemble` and restart the docker containers):
```
--- a/airbyte-server/src/main/java/io/airbyte/server/apis/SourceApiController.java
+++ b/airbyte-server/src/main/java/io/airbyte/server/apis/SourceApiController.java
@@ -57,7 +57,11 @@ public class SourceApiController implements SourceApi {
   @Secured({EDITOR})
   @Override
   public CheckConnectionRead checkConnectionToSourceForUpdate(final SourceUpdate sourceUpdate) {
-    return ApiHelper.execute(() -> schedulerHandler.checkSourceConnectionFromSourceIdForUpdate(sourceUpdate));
+    return ApiHelper.execute(() -> {
+      final CheckConnectionRead r = schedulerHandler.checkSourceConnectionFromSourceIdForUpdate(sourceUpdate);
+      r.setJobInfo(r.getJobInfo().connectorConfigurationUpdated(true));
+      return r;
+    });
   }

   @Post("/clone")
@@ -95,7 +99,11 @@ public class SourceApiController implements SourceApi {
   @Secured({READER})
   @Override
   public SourceRead getSource(final SourceIdRequestBody sourceIdRequestBody) {
-    return ApiHelper.execute(() -> sourceHandler.getSource(sourceIdRequestBody));
+    return ApiHelper.execute(() -> {
+      final SourceRead r = sourceHandler.getSource(sourceIdRequestBody);
+      r.setName(r.getName() + System.currentTimeMillis());
+      return r;
+    });
   }
```

`checkConnectionToSourceForUpdate` automatically sets the flag to true for all connection checks. The `getSource` endpoint adds the current time in ms to the name of the connection to make it obvious the form actually updated.

https://www.loom.com/share/7d73b667dc4b40378f268ad8bf987f22

### CODE REVIEW

File: airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx [hunk 1]
```diff
@@ -129,7 +131,7 @@ export const ConnectorCard: React.FC<ConnectorCardCreateProps | ConnectorCardEdi
   const testConnectorWithTracking = async (connectorCardValues?: ConnectorCardValues) => {
     trackTestConnectorStarted(selectedConnectorDefinition);
     try {
-      await testConnector(connectorCardValues);
+      return await testConnector(connectorCardValues);
       trackTestConnectorSuccess(selectedConnectorDefinition);
```
Comment (lmossman) [inline airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx:135]: If we `return` here on line 134, doesn't that mean we will never execute the next line to track the test success?

It feels like we should instead be saving the result of `testConnector()` to a variable, then calling track, then returning that variable

**Review by lmossman — APPROVED** [review]
I tested this out locally following your instructions, and I noticed some weird behavior which I think is only happening because the test isn't matching prod behavior, but I wanted to double check my understanding:
- I tried editing the `Pokemon Name` on one of my existing connectors, and when I clicked `Test and save`, the `Source name` was updated with the current timestamp as expected, but the `Pokemon Name` was set back to the name that had been saved to the source before I changed it.
- I believe this is expected from this test though, because we just modified the airbyte-server to return a response with the `connectorConfigurationUpdated` flag set to true, and only the connector name is changed in the response of `getSource()`, so the new pokemon name was never actually saved.
- In the production use case that this PR is handling, the `checkConnectionToSourceForUpdate()` call would actually update the source to have the new pokemon name, so the subsequent `getSource()` call would respond with that new name, and it therefore wouldn't go back to the old name.

If the above sounds correct to you, then I think the changes here look good. 

I just had one question about a `return` that you added, but I think it can be fixed without me needing to give it another look

File: airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx (hunk 1 above)
Comment (flash1293) [reply airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx]: Good catch, no idea how I missed this. @timroes is there a reason not to fail the build on unreachable code?

**Comment by flash1293** [comment]
> If the above sounds correct to you, then I think the changes here look good.

Exactly, that's how I understood the flow. Just to make sure - @pedroslopez can you confirm that if `jobInfo.connectorConfigurationUpdated` is set, then the whole config object that was sent to the  `check_for_update` endpoint is persisted already, right?

**Comment by pedroslopez** [comment]
> > If the above sounds correct to you, then I think the changes here look good.
> 
> Exactly, that's how I understood the flow. Just to make sure - @pedroslopez can you confirm that if `jobInfo.connectorConfigurationUpdated` is set, then the whole config object that was sent to the `check_for_update` endpoint is persisted already, right?

Yep! With the caveat that it’s likely not the exact config object sent as part of the request, since this happens because the connector itself wanted to change some config values.

**Pull request merged**